### PR TITLE
Escape ld+json content with manual JSON builder.

### LINKED ISSUE #5402: Error with encoding linked data for packages with double quotes in their descriptions
image

### PR DESCRIPTION
**Pull request by isoos**
- Fixes #5402.
- Using `json` encoder + second encoder doesn't work, as we want to double-encode some characters, but not the escape `\`. Detecting it correctly seems to need implementing a proper parser to do it right.
- Instead, this code emits the JSON structure `{`..`}`, `[`..`]`, `key: value` pairs directly, and also encodes the strings with unicode escaping for almost all of the characters, making sure that both the JSON and the HTML structures are escaped.

### CODE REVIEW

File: app/lib/frontend/dom/dom.dart [hunk 1]
```diff
@@ -135,17 +134,49 @@ Node codeSnippet({
...
+        if (i > 0) sb.write(',');
+        write(entries[i].key);
+        sb.write(':');
+        write(entries[i].value);
+      }
+      sb.write('}');
+    } else if (value is bool || value is num || value == null) {
+      sb.write(json.encode(value));
+    } else {
+      throw ArgumentError('Unexpected value: $value');
```
Comment (jonasfj) [inline app/lib/frontend/dom/dom.dart:172]: Better explain that the `content` passed to `ldJson` could not be translated to JSON, because it contains a value of type `value.runtimeType` (or something like that)

File: app/lib/frontend/dom/dom.dart [hunk 2]
```diff
@@ -16,8 +16,7 @@ final _elementRegExp = _attributeRegExp;
 // As we want to store rawJson inside a HTML Element, it is better
 // to escape all non-trusteded characters inside it. Non-trusted
 // characters must include `</!>` characters.
-final _ldJsonEscapedCharactersRegExp =
-    RegExp(r'[^0-9a-zA-Z ,@\:\{\}\[\]\.\-"]');
+final _ldJsonEscapedCharactersRegExp = RegExp(r'[^0-9a-zA-Z ,@\.\-]');
```
Comment (jonasfj) [inline app/lib/frontend/dom/dom.dart:19]: Why this change?

I get that we need to escape `"`, but why any of the others?

**Review by jonasfj — APPROVED** [review]

File: app/lib/frontend/dom/dom.dart (hunk 2 above)
Comment (jonasfj) [reply app/lib/frontend/dom/dom.dart:19]: Just thinking it's nice that URLs can be read even if escaping somehow fails...

File: app/test/frontend/golden/landing_page.html [hunk 3]
```diff
@@ -221,6 +221,6 @@ <h1 class="home-block-title">Package of the Week</h1>
         <img class="inline-icon" src="/static/img/bug-report-white-96px.png?hash=mocked_hash_709947442" alt="bug report icon" width="20" height="20" title="Report an issue with this site"/>
       </a>
     </footer>
-    <script type="application/ld+json">{"@context":"http:\u002f\u002fschema.org","@type":"WebSite","url":"https:\u002f\u002fpub.dev\u002f","potentialAction":{"@type":"SearchAction","target":"https:\u002f\u002fpub.dev\u002fpackages\u003fq\u003d{search\u005fterm\u005fstring}","query-input":"required name\u003dsearch\u005fterm\u005fstring"}}</script>
+    <script type="application/ld+json">{"@context":"http\u003a\u002f\u002fschema.org","@type":"WebSite","url":"https\u003a\u002f\u002fpub.dev\u002f","potentialAction":{"@type":"SearchAction","target":"https\u003a\u002f\u002fpub.dev\u002fpackages\u003fq\u003d\u007bsearch\u005fterm\u005fstring\u007d","query-input":"required name\u003dsearch\u005fterm\u005fstring"}}</script>
```
Comment (jonasfj) [inline app/test/frontend/golden/landing_page.html:224]: @isoos are you sure we want to escape `:` in strings?

Comment (isoos) [reply app/test/frontend/golden/landing_page.html:224]: I don't mind escaping them. We won't have the `//` parts either, so URLs will be escaped, it is unlikely that they will be nice to read unparsed.
